File: tests/solutions_cpp/binary_tree_maximum_path_sum.cpp
#include "json_helper.h"

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        best = INT_MIN;
        dfs(root);
        return best;
    }
private:
    int best;
    int dfs(TreeNode* node) {
        if (!node) return 0;
        int l = max(0, dfs(node->left));
        int r = max(0, dfs(node->right));
        best = max(best, node->val + l + r);
        return node->val + max(l, r);
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto root = build_tree(args[0]);
    cout << sol.maxPathSum(root) << endl;
    return 0;
}

File: tests/solutions_js/binary_tree_maximum_path_sum.js
class Solution {
    maxPathSum(root) {
        this.maxSum = -Infinity;

        const dfs = (node) => {
            if (node === null) return 0;
            const leftGain = Math.max(dfs(node.left), 0);
            const rightGain = Math.max(dfs(node.right), 0);
            this.maxSum = Math.max(this.maxSum, node.val + leftGain + rightGain);
            return node.val + Math.max(leftGain, rightGain);
        };

        dfs(root);
        return this.maxSum;
    }
}
